From 6f9835c187b7961edb4dd0203b615589f9023b8c Mon Sep 17 00:00:00 2001
From: "Sv. Lockal" <lockalsash@gmail.com>
Date: Mon, 6 Jan 2025 06:53:06 +0000
Subject: [PATCH] Fix compilation with libc++

`std::vector<char[16]>` works with libstdc++, but fails with libc++.
The issue was reported to https://github.com/llvm/llvm-project/issues/98734, and according to discussion, is not a part of c++17 (used currently in hipBLASLt).
This change switches `char[16]` to `std::array<char, 16>`, which is compliant with c++17.
---
 .../src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp  | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/library/src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp b/library/src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp
index 9ed4053b78..9e4c2bbe1b 100644
--- a/library/src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp
+++ b/library/src/amd_detail/rocblaslt/src/rocblaslt_mat.cpp
@@ -28,6 +28,7 @@
 #include "handle.h"
 #include "rocblaslt_mat_utils.hpp"
 #include "tensile_host.hpp"
+#include <array>
 
 #include <hip/hip_runtime_api.h>
 
@@ -393,7 +394,7 @@ rocblaslt_status
     std::vector<int64_t>            ldc_vec, batch_stride_c_vec, num_batches_c_vec;
     std::vector<int64_t>            ldd_vec, batch_stride_d_vec, num_batches_d_vec;
     std::vector<int64_t>            lde_vec, batch_stride_e_vec, num_batches_e_vec;
-    std::vector<int8_t[16]>         alpha_1(matmul_descr.size());
+    std::vector<std::array<int8_t, 16>>         alpha_1(matmul_descr.size());
 
     std::vector<bool> gradient_vec;
 
@@ -493,10 +494,10 @@ rocblaslt_status
             return validArgs;
 
         const void* alphaTmp = nullptr;
-        memset(alpha_1[i], 0, sizeof(int8_t) * 16);
+        memset(alpha_1[i].data(), 0, sizeof(int8_t) * 16);
         if(scaleAlphaVec)
         {
-            setTo1(compute_type, (void*)alpha_1[i], &alphaTmp);
+            setTo1(compute_type, (void*)alpha_1[i].data(), &alphaTmp);
         }
         else
         {
@@ -1193,7 +1194,7 @@ rocblaslt_status rocblaslt_groupedgemm_create_cpp_impl_2(const rocblaslt_handle
     std::vector<int64_t> lde_vec, batch_stride_e_vec, num_batches_e_vec;
     std::vector<bool>    gradient_vec;
 
-    std::vector<int8_t[16]> alpha_1(m.size());
+    std::vector<std::array<int8_t, 16>> alpha_1(m.size());
 
     for(int i = 0; i < m.size(); i++)
     {
@@ -1287,10 +1288,10 @@ rocblaslt_status rocblaslt_groupedgemm_create_cpp_impl_2(const rocblaslt_handle
             return validArgs;
 
         const void* alphaTmp = nullptr;
-        memset(alpha_1[i], 0, sizeof(int8_t) * 16);
+        memset(alpha_1[i].data(), 0, sizeof(int8_t) * 16);
         if(scaleAlphaVec)
         {
-            setTo1(compute_type, (void*)alpha_1[i], &alphaTmp);
+            setTo1(compute_type, (void*)alpha_1[i].data(), &alphaTmp);
         }
         else
         {
